### PR TITLE
make composer work with multiple different subscriptions

### PR DESCRIPTION
**Pull request by msehnout**
Current status:
 * ready for review

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [x] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [X] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

Related:
https://github.com/osbuild/guides/pull/36
osbuild/osbuild#645

### CODE REVIEW

File: test/cases/satellite.sh
Comment (ondrejbudai) [inline test/cases/satellite.sh:73]: This should go to the spec file. :)

File: test/cases/satellite.sh
Comment (atodorov) [inline test/cases/satellite.sh:66]: Doesn't httpd come with a working config file? What's the difference here ?

File: test/cases/satellite.sh
Comment (atodorov) [inline test/cases/satellite.sh:143]: I think you should be able to drop this under `/etc/httpd/conf.d/` instead of overriding httpd.conf

File: test/cases/satellite.sh
Comment (atodorov) [inline test/cases/satellite.sh:220]: missing newline at EOF.

File: test/cases/satellite.sh
Comment (atodorov) [inline test/cases/satellite.sh:200]: NOTE: this override will potentially conflict with other internal overrides that we do, especially when this test is enabled on RHEL, see #1318. 

This test script should be executed on a stand-alone runner so that it doesn't affect other executions afterwards.

File: schutzbot/Jenkinsfile
Comment (atodorov) [inline schutzbot/Jenkinsfile:820]: This also needs to be added into GitLab but is currently blocked by #1408 !  Also see https://github.com/osbuild/osbuild-composer/pull/1405#discussion_r630822032

**Comment by msehnout** [comment]
Depends on: https://github.com/osbuild/osbuild/pull/645

File: internal/distro/fedora32/distro.go
Comment (ondrejbudai) [inline internal/distro/fedora32/distro.go:20]: I think we should reference `releaseVersion` in the other constants (`defaultName`, `modulePlatformID`, `ostreeRef`). Something for a follow-up.

File: internal/rpmmd/repository.go
Comment (ondrejbudai) [inline internal/rpmmd/repository.go:372]: We should at least log the error. Hiding it makes debugging difficult. Also `s/subscptions/subscriptions`. :)

File: internal/distro/fedora32/distro.go
Comment (ondrejbudai) [inline internal/distro/fedora32/distro.go:270]: Is there any usecase for Fedora and rhsm?

File: test/cases/regression-composer-works-behind-satellite.sh
Comment (ondrejbudai) [inline test/cases/regression-composer-works-behind-satellite.sh:186]: Just so we don't forget: This needs to go away before we merge this.

**Review by ondrejbudai — COMMENTED** [review]
Awesome stuff, thanks! Looking forward to shipping this!

File: internal/rhsm/secrets.go
Comment (teg) [inline internal/rhsm/secrets.go:35]: Is the special semantics of this file documented anywhere we could refer to? Might make sense to explain in some detail. Could people use other repo files for the same purpose, or is this special?

File: internal/rpmmd/repository.go
Comment (teg) [reply internal/rpmmd/repository.go:372]: We should probably treat the cases of no subscriptions gracefully and not return an error in that case (it is valid to have an unsubscribed system), and then only fail on real errors here.

We should, however, fail (with a helpful error code) in case someone tries to do something later on that requires subscriptions and we have none.

**Review by teg — COMMENTED** [review]
Love this! Only very small nit picks/questions inline.

File: test/cases/regression-composer-works-behind-satellite.sh
Comment (msehnout) [reply test/cases/regression-composer-works-behind-satellite.sh:186]: Sure, I hope osbuild gets a new release soon and I can remove this.

File: internal/distro/fedora32/distro.go
Comment (msehnout) [reply internal/distro/fedora32/distro.go:270]: Very unlikely, but it should be possible to setup private mirror (e.g. using Satellite or Pulp) with Fedora content and require subscription. Anyway it is useful for the test case I introduced.

File: internal/rpmmd/repository.go
Comment (msehnout) [reply internal/rpmmd/repository.go]: I implemented these error handling cases:
 1. The repo file is not there - print a message about it
 2. The repo file is there but it is not possible to open it or parse it - print a different message
 3. User attempts to use rhsm on a system with no subscriptions - suggest that the system must be subscribed

File: internal/rhsm/secrets.go
Comment (msehnout) [reply internal/rhsm/secrets.go:36]: It is not different from any other repo file. I wrote this fact into a comment.

Comment (teg) [reply internal/rhsm/secrets.go:39]: So the reason that it is special is that it is the file written by RHSM, right? I.e., it does not make sense for us to support the same logic for other repo files, as we are explicitly targeting RHSM?

**Review by teg — APPROVED** [review]
osbuild has been released, so this is ready to go in (after dropping the repo pinning from Schutzfile). Great work!

**Review by teg — DISMISSED** [review]
This is great! Thanks for all the work you did on this!

File: test/cases/satellite.sh
Comment (msehnout) [reply test/cases/satellite.sh]: I've created a cleanup code at the end of the script. Do you think we should run it in a stand-alone machine anyway?

File: schutzbot/Jenkinsfile
Comment (msehnout) [reply schutzbot/Jenkinsfile]: This is now outdated. The test case is executed from the regression.sh script which in turn is executed from Gitlab CI configuration.

File: test/cases/regression.sh
Comment (atodorov) [inline test/cases/regression.sh:51]: Can this test script be made more generic and also work on RHEL ?

**Review by atodorov — COMMENTED** [review]
Added questions.

File: test/cases/regression.sh
Comment (msehnout) [reply test/cases/regression.sh:51]: It should be possible, the problem is that the script creates a proxy for RPMrepo which works in a different way for RHEL and Fedora. Fedora repositories are hosted on AWS, but RHEL repos are hosted on PSI. I'm not sure we can make this change in time for the Z stream update.

**Review by atodorov — DISMISSED** [review]

File: internal/rpmmd/repository.go
Comment (ondrejbudai) [inline internal/rpmmd/repository.go:380]: typo

File: test/cases/regression-composer-works-behind-satellite.sh
Comment (ondrejbudai) [inline test/cases/regression-composer-works-behind-satellite.sh:66]: I suggest comparing only the major version against `${VERSION_ID%.*}`.

File: internal/rpmmd/repository.go
Comment (msehnout) [reply internal/rpmmd/repository.go]: Fixed, thanks 👍

File: test/cases/regression-composer-works-behind-satellite.sh
Comment (msehnout) [reply test/cases/regression-composer-works-behind-satellite.sh]: Fixed, thanks 👍

**Review by teg — DISMISSED** [review]
Thanks so much for getting this over the finishing line, this will be very much appreciated :)

**Review by ondrejbudai — DISMISSED** [review]

**Comment by ondrejbudai** [comment]
I rebased this PR to fix the conflict in `go.sum`.

**Review by teg — APPROVED** [review]

**Review by ondrejbudai — APPROVED** [review]

**Pull request merged**
